core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -155,19 +181,37 @@
                 }
                 curr = curr.next;
             }
-            return false;
+            return hasAttrInParent(key);
         }
     }
 
-    /**
-     * Returns all {@link Attribute}s this map contains.
-     */
+    private <T> boolean hasAttrInParent(AttributeKey<T> key) {
+        if (parentMap != null) {
+            // This happens at least only once, because the parentMap does not have its parent.
+            return parentMap.hasAttr(key);
+        }
+        return false;
+    }
+
+    @Override
     public Iterator<Attribute<?>> attrs() {
+        return attrs(false);
+    }
+
+    public Iterator<Attribute<?>> attrs(boolean ownAttrs) {
         final AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
         if (attributes == null) {
+            if (!ownAttrs && parentMap != null) {
+                // This happens at least only once, because the parentMap does not have its parent.
+                return parentMap.attrs();
+            }
             return Collections.emptyIterator();
         }
 
+        if (!ownAttrs && parentMap != null) {
+            return Iterators.concat(new IteratorImpl(attributes), parentMap.attrs());
```

Comment:
You can't just concat. You have to skip the parent attributes if they exist in own attributes.

Reply:
Couldn't think of it. Thanks!